README.md
# DSH enhanced version
 - [DSH](https://www.netfort.gr.jp/~dancer/software/dsh.html.en): dancer's shell / distributed shell, 

# Change list:
 - 1.change default shell from rsh to ssh
 - 2.add command parameter to provide login user name: -u | --user username 
   - this is very important for me. So I can change remote user without modify the group conifgruation file or ssh_config file on kerberos server. 
   - Login user are considered in the following order:
     - 1.new added param: -u|--user username
     - 2.username@machinename in dsh configuration file(group or machines)
     - 3.ssh_config file(~/.ssh/config)
     - 4.current username

# Base Version:
 - dsh-0.22: https://www.netfort.gr.jp/~dancer/software/downloads/dsh-0.22.0.tar.gz
 - libdshconfig-0.20.10.cvs.1: https://www.netfort.gr.jp/~dancer/software/downloads/libdshconfig-0.20.10.cvs.1.tar.gz
 
# Build & Configuration

 - 1.build(consider you are in kerberos server, can not install to default position)
```
cd libdshconfig
./configure --prefix=$HOME/local && make && make install

cd ../dsh
./configure --prefix=$HOME/local \
  LDFLAGS=-L$HOME/local/lib \
  CPPFLAGS=-I$HOME/local/include \
  && make \
  && make install
```
 - 2.add ~/local/bin to your PATH environment variable.


 
 


libdshconfig/NEWS
This is an empty news file.

libdshconfig/tests/test-basic.sh
#! /bin/sh
# Basic testing program for dshconfig
# read a config file, and output the file processed.

set -ex

./test-dshconfig < ${srcdir}/tests/test-basic.input > tests-basic.o
diff ${srcdir}/tests/test-basic.output tests-basic.o
rm -f tests-basic.o


libdshconfig/tests/test-longline.sh
#! /bin/sh
# try using some perl to run this test.
set -e
perl -e 'print ("x"x299900 . ": " . "y"x299900 . "\n")x20 ' | ./test-dshconfig > /dev/null

libdshconfig/tests/test-dshconfig-searchdata.sh
#! /bin/sh
# test-run the application.

READINGSOURCE=${srcdir}/tests/test-dshconfig-searchdata.data ./test-dshconfig-searchdata


dsh/dsh.c
/*
 *  DSH / dancer's shell or the distributed shell
 *  Copyright (C) 2001, 2002 Junichi Uekawa
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Main program. 
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <locale.h>

#include "dsh.h"
#include "linkedlist.h"
#include "parameter.h"

#include "gettext.h"
#define _(A) gettext(A)

char * remoteshell_command="ssh";
char * login_user=NULL; /* TODO: not free? exit after exec, not problem */

int verbose_flag=0;		/* verbosity flag */
int wait_shell=1;		/* waiting for shell to execute (concurrence) */
int pipe_option=0;	        /* show machine names -- piping option. */
int num_topology=1;		/* number of topology to use as a block to execute rsh. 
				 1 = for-loop
				 2 = binary-tree
				 4 = quad-tree.
				*/
linkedlist* remoteshell_command_opt_r=NULL; /* reverse-ordered list of rsh options. */
int buffer_size = 1024;		/* buffer size. */






/* function defining "getline" */
#ifndef HAVE_GETLINE
/* an imcomplete, and wrong implementation of getline */
ssize_t getline (char **LINEPTR, size_t *N, FILE *STREAM)
{
  const int GETLINESIZE = 256;
  int fgl;
  
  if (!*LINEPTR)
    *LINEPTR= malloc (GETLINESIZE);
  if (*N != GETLINESIZE)
    *LINEPTR = realloc (*LINEPTR, GETLINESIZE);
  if (!*LINEPTR)
    {
      return -1;
    }
  if (!fgets (*LINEPTR, GETLINESIZE - 1, STREAM))
    return -1;
  
  *N = strlen (*LINEPTR);
  return GETLINESIZE;
  
}
#endif

/* define asprintf if it doesn't exist */
#ifndef HAVE_ASPRINTF
#include <stdarg.h>
int asprintf(char **strp, const char *fmt, ...)
{
  ssize_t buflen = 50 * strlen(fmt); /* pick a number, any number */
  *strp = malloc(buflen);

  if (*strp)
  {
    va_list ap;
    va_start(ap, fmt);
    vsnprintf(*strp, buflen, fmt, ap);
    va_end(ap);
    return buflen;
  } 
  return -1;
}
#endif /* HAVE_ASPRINTF  */

#if !defined(WAIT_ANY)
#define WAIT_ANY ((pid_t)-1)
#endif /* WAIT_ANY */

/**
 *  A process that just echoes back, reading fd_in using getline, 
 *  and writing to fd_out
 *
 * @returns -1 on error, 0 on success.
 */
static int
do_echoing_back_process(int fd_in, int fd_out, const char * prompt)
{
  char * buf = NULL;
  int bufsize = 0;
  FILE*f = fdopen (fd_in, "r");
  FILE*standard_output = fdopen (fd_out, "w");
  
  if (!f || !standard_output)
    {
      fprintf(stderr, "%s: Could not open descriptor [%i] or [%i]\n", PACKAGE, fd_in, fd_out);
      return -1; 
    }
  
  while (0 <  getline(&buf, &bufsize, f))
    {
      fprintf (standard_output, "%s: %s", prompt, buf);
    } 
  if (buf)
    free (buf);

  return 0;
}

/**
 * Forks the output pipe routine for fd[x]
 *
 * @returns -1 on error, 0 on success
 */
static int 
fork_and_pipe_echoing_routine (
			       /** File descriptor ID */ int fdid,
			       const char * machinename) 
{
  int childid;
  int capture_fd[2];
  if (-1 == pipe (capture_fd))
    {
      fprintf(stderr, _("%s: cannot create pipe\n"), PACKAGE);
      return -1;
    }
  if (0 != (childid = fork()))
    {			
      /* The child process.  */
      close (capture_fd[1]);
      if (do_echoing_back_process(capture_fd[0], fdid, machinename))
	exit (EXIT_FAILURE);
      exit(EXIT_SUCCESS);	/* actually, it should just get the
				   child process return code. */
    }
  else if (childid == -1)
    {			/* do some error processing */
      fprintf (stderr, _("%s: Cannot spawn process\n"), PACKAGE);
      return -1;
    }      
  else
    {
      if (-1 == dup2 (capture_fd[1], fdid))
	{
	  fprintf(stderr, PACKAGE ": %s\n", _("Failed playing with pipe"));
	  return -1;
	}
      close (capture_fd[0]);
      close (capture_fd[1]);
    }
  return 0;
}

/** 
 * Actually execute the program, with maybe a pipe process being forked.
 * the actual program execution is done by llexec.
 * This part does the optional pipe construction or direct llexec
 * depending on the flags.
 *
 * @returns -1 on failure
 */
static int
do_execute_with_optional_pipe (const char * remoteshell_command,
			       const linkedlist * commandline,
			       int pipe_option , /** 1=pipe output and put
						  pretty flags*/
			       const char * machinename)
{
  if (pipe_option & PIPE_OPTION_OUTPUT)		/* pipe the outputs */
    {
      if (fork_and_pipe_echoing_routine(1, machinename) ||
	  fork_and_pipe_echoing_routine(2, machinename))
	{	  
	  fprintf(stderr, _("%s: Failed on constructing a pipe and forking\n"), PACKAGE);
	  return -1;
	}
    }

  llexec ( remoteshell_command, commandline );
  fprintf(stderr, _("%s: Failed executing %s with llexec call\n"), PACKAGE, remoteshell_command);
  exit (EXIT_FAILURE);
}



static int * fd_output_array = NULL; /** array of fd to duplicate input to */
static int fd_output_array_len = 0;
/**
 * adds an fd to the list of fds to be processed by the input-duplication daemon.
 */
static void
add_fd_to_output_array(int fd)
{
  fd_output_array_len ++ ;
  fd_output_array = realloc (fd_output_array, fd_output_array_len * sizeof(int));
  fd_output_array[fd_output_array_len - 1] = fd;
}

/**
 * spawns rsh/ssh session on single machine
 *
 * @returns -1 on failure, 0 on success
 */
static int
execute_rsh_single (const char * remoteshell_command, 
		    const linkedlist * remoteshell_command_opt_r,
		    const char * param_username,
		    const char * param_machinename,
		    const linkedlist * rshcommandline_r,
		    int pipe_option /** The pipe option */)
{  
  int childpid;
  int childstatus;
  int input_pipe [2];
  
  if (pipe_option & PIPE_OPTION_INPUT)
    {
      pipe (input_pipe);
    }  

  /* Execute the rsh process */
  if (0==(childpid=fork()))
    {				/* child process */
      /* Must NOT fork again from this point until I say otherwise */
      linkedlist * tmp = NULL;
      char * full_machinename = strdup (param_machinename); /* we will always exit() from here, so
				       no need to free this myself. */
      char * machinename = full_machinename; 

      linkedlist * local_remoteshell_command_opt_r = lldup(remoteshell_command_opt_r);      

      /* input piping */
      if (pipe_option & PIPE_OPTION_INPUT)
	{
	  dup2 (input_pipe[0],0);
	  close (input_pipe[1]);
	  close (input_pipe[0]);
	  /* Must NOT fork again before this point. */
	}

	char * username = NULL;
	if(param_username != NULL){
	    /* if provide username */
           username = param_username;
           if (NULL != (machinename = strchr(full_machinename,'@')))
            {			/* username was specified */
               *(machinename++) = 0 ; /* cut the string at @ and point to following string (machine name) */
            } else {
               machinename = full_machinename;
            }
	} else {
        username = machinename;
        /* process to handle username@hostname */
        if (NULL != (machinename = strchr(machinename,'@')))
        {			/* username was specified */
            *(machinename++) = 0 ; /* cut the string at @ and point to following string (machine name) */
        }
        else
        {			/* no username was specified */
            machinename=username;
            username=NULL;
        }
    }

    if(username != NULL){
        local_remoteshell_command_opt_r = lladd (local_remoteshell_command_opt_r, "-l");
        local_remoteshell_command_opt_r = lladd (local_remoteshell_command_opt_r, username);
        username = NULL;
    }

      tmp = llcat (tmp, local_remoteshell_command_opt_r);
      tmp = lladd (tmp, machinename);
      tmp = llcat (tmp, lldup(rshcommandline_r));
      tmp = llreverse(tmp);
      
      if (verbose_flag) 	/* debugging */
	{
	  printf(_( "DUMPing parameters passed to llexec\n"));
	  lldump(tmp);
	}	  
      
      if (-1 == (do_execute_with_optional_pipe(remoteshell_command, 
					       tmp, pipe_option, 
					       param_machinename)))
	{
	  fprintf(stderr,  
		  _("%s: Failed to execute remote shell command %s\n"), 
		  PACKAGE, remoteshell_command);
	  exit (EXIT_FAILURE);
	}
				/* what follows should never occur */
      fprintf (stderr, 
	       _("%s: Unexpected error occurred, do_execute_with_optional_pipe failed, and returned an error code that is not -1\n"), PACKAGE);
      exit (EXIT_FAILURE);
    }
  else if (childpid==-1)
    {
      /* fork failed */
      fprintf (stderr, 
	       _("%s: fork failed, in execute_rsh_single\n"), PACKAGE);
      return -1;
    }
  else
    {
      if (pipe_option & PIPE_OPTION_INPUT)
	{
	  close (input_pipe[0]);
	  /* add input_pipe[1] to the array of outputs one 
	     must handle here... */
	  add_fd_to_output_array(input_pipe[1]);
	}
      if (wait_shell)
	  waitpid(childpid, &childstatus, 0);	/* wait for termination, 
						   if it was required */
      return 0;
    }
  /* nobody reaches here. */
}


/**
 * spawns dsh session on other machines
 *
 * TODO: quote properly.
 */
static int
execute_rsh_multiple (const char * remoteshell_command,
		      const linkedlist * remoteshell_command_opt_r,
		      const char* username_r,
		      const linkedlist * machinelist, 
		      int nummachines, 
		      const linkedlist * rshcommandline_r)
{
  linkedlist* extraparam = NULL;
  linkedlist* tmp;  
  linkedlist* tmp_start ;  
  char * buffer;

  extraparam = lladd(extraparam, DSH_COMMANDLINE);

  tmp = tmp_start = lldup(machinelist);  
  while (tmp && (nummachines -- > 0))
    {
      extraparam=lladd (extraparam, "-m");
      extraparam=lladd (extraparam, tmp->string);
      tmp=tmp->next;
    }
  llfree (tmp_start);  
  tmp = tmp_start = llreverse(lldup(remoteshell_command_opt_r));
  while (tmp)
    {
      extraparam=lladd (extraparam, "-o");
      extraparam=lladd (extraparam, tmp->string);
      tmp=tmp->next;
    }
  llfree(tmp_start);  
				/* hand over any extra options here. */
  if (verbose_flag)
    extraparam=lladd (extraparam, "-v");
  if (wait_shell)
    extraparam=lladd (extraparam, "-w");
  else
    extraparam=lladd (extraparam, "-c");
  if (pipe_option)
    extraparam=lladd (extraparam, "-M");
  

  asprintf(&buffer, "-n%i", num_topology);  
  extraparam=lladd (extraparam, buffer);
  free(buffer);  

  /* add remoteshell command */
  asprintf(&buffer, "-r%s", remoteshell_command);  
  extraparam=lladd (extraparam, buffer);
  free(buffer);  

				/* end of extra options. */
  extraparam=lladd (extraparam, "--");
  rshcommandline_r = llcat (extraparam, lldup(rshcommandline_r));

  return execute_rsh_single (remoteshell_command, remoteshell_command_opt_r, username_r, machinelist->string, rshcommandline_r, 0);
}

/**
   execute remote shell 
   executes dsh on other machines, or
   executes rsh to execute command.
   depending on the parameter and 
   status.

   @return 0 on success, -1 on failure.
*/
static int
execute_rsh ( const char * remoteshell_command, 
	      const linkedlist * remoteshell_command_opt_r,
	      const char* username_r,
	      const linkedlist * machinelist,
	      int nummachines /** The number of machines to invoke rsh at the same time */,
	      const linkedlist * rshcommandline_r)
{				
  if (nummachines == 1)
    return execute_rsh_single (remoteshell_command, remoteshell_command_opt_r, username_r, machinelist->string, rshcommandline_r, pipe_option);
  else
    return execute_rsh_multiple (remoteshell_command, remoteshell_command_opt_r, username_r, machinelist, nummachines, rshcommandline_r);
}

/**
 * Forks off to do read from input, and duplicate it to
 * output the same thing to all of individual remote processes
 */
void
run_input_forking_child_processes_process()
{
  char * buf;
  int count ;
  int i;
  
  switch (fork())
    {
    case 0:
      /* the child process */
      buf = malloc_with_error ( buffer_size );
      while ((count = read(0, buf, buffer_size)) != -1 )
	{
	  if (count == 0)	/* if there is zero-byte read, it is an end-of-file */
	    break;
	  for (i = 0; i < fd_output_array_len; ++i )
	    {
	      write (fd_output_array [i], buf, count);
	    }
	}      
      for (i = 0; i < fd_output_array_len; ++i )
	{
	  close (fd_output_array [i]);
	}
      exit (EXIT_SUCCESS);
    case -1:
      /* fork failed */
      break;
    default:
      /* parent process closes the output array. */
      for (i = 0; i < fd_output_array_len; ++i )
	{
	  close (fd_output_array [i]);
	}
      if (verbose_flag)
	printf (_("%s: forked off input forking process\n"), PACKAGE);
      /* do not read from stdin. */
      open_devnull();
    }
}


/**
   Called after the command-line parsing.
   do the shell execution without caring
   about what actually would happen

   @returns -1 on failure, 0 on success
*/
int
do_shell (linkedlist* machinelist, linkedlist*rshcommandline_r)
{
  int nummachines = llcount(machinelist) / num_topology ;  
  int i;

				/* topology 1 is special. */
  if ( nummachines == 0 || (num_topology == 1))
    {
      nummachines = 1;
    }    
  
  while (machinelist)
    {
      if (wait_shell && verbose_flag) 
	{
	  fprintf (stderr, _("--- Executing on %s \n"), machinelist->string);
	}
      if (execute_rsh (remoteshell_command, remoteshell_command_opt_r, login_user, machinelist, nummachines, rshcommandline_r))
	{
	  fprintf(stderr, _("%s: execute_rsh failed, rsh invocation failure.\n"), PACKAGE);
	  return -1;
	}
      
      /* skip the machines that program was executed on, and repeat */
      for (i=0; (i < nummachines) && machinelist; ++i)
	  machinelist = machinelist -> next;
    }

  /* I have executed on all processes, I can now start cleaning up process...  */
  if (pipe_option &= PIPE_OPTION_INPUT)
    run_input_forking_child_processes_process();  

  if (!wait_shell)
    while(-1 != (waitpid(WAIT_ANY, NULL, 0)));	/* waiting for all. */
  
  if (verbose_flag)
    fprintf(stderr, _("--- Terminated running\n"));
  
  return 0;  
}


int
main(int ac, char ** av)
{
  char *buf=NULL;
  
  setlocale (LC_ALL, "");
  if (!textdomain(PACKAGE_NAME))
    {
      if (!bindtextdomain(PACKAGE_NAME, LOCALEDIR))
	fprintf (stderr, "%s: failed to call bindtextdomain\n", PACKAGE);
    }
  
  
  
  load_configfile(DSH_CONF);
  if (asprintf (&buf, "%s/.dsh/dsh.conf", getenv("HOME")) < 0)
    {
      fprintf (stderr, _("%s: asprintf failed\n"), PACKAGE);
      exit (1);
    }  
  load_configfile(buf);
  free (buf);

  return parse_options(ac, av);  
}


dsh/parameter.h
/*
 *  DSH / dancer's shell or the distributed shell
 *  Copyright (C) 2001, 2002 Junichi Uekawa
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

void * malloc_with_error(int size);
int load_configfile(const char * dsh_conf);
int parse_options ( int ac, char ** av);
extern char * remoteshell_command;
extern char * login_user;

extern int  verbose_flag;		/** verbosity flag */
extern int wait_shell;		/** waiting for shell to execute (concurrence) */
extern int  pipe_option;	/** &1 = show machine names ...  
				   &2 = duplicate input.
				 */
extern int  num_topology;		/** number of topology to use as a block to execute rsh. 
				 1 = for-loop
				 2 = binary-tree
				 4 = quad-tree.
				*/
extern linkedlist* remoteshell_command_opt_r; /** reverse-ordered list of rsh options. */
extern int buffer_size;		/* buffer size for running the input process. */
void open_devnull(void);

/* function defined in dsh.c */
int do_shell(linkedlist * machinelist, linkedlist*);



dsh/parameter.c
/*
 *  DSH / dancer's shell or the distributed shell
 *  Copyright (C) 2001, 2002 Junichi Uekawa
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Command-line and parameter handling routines.
 */
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <netdb.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/wait.h>

#include "dsh.h"

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#include "config.h"
#include "linkedlist.h"
#include "parameter.h"
#include "libdshconfig.h"

#include "gettext.h"
#define _(A) gettext(A)

/**
 * allocate memory with error
 *
 * @return allocated memory
 */
void * 
malloc_with_error(int size)
{
  void * u = malloc (size);
  if (!u)
    {
      fprintf (stderr, _("%s: failed to allocate memory of %i bytes\n"), PACKAGE, size);
      exit(1);      
    }
  return u;  
}

/* 
 * remove \n and other things from string
 * and return the new pointer.
 */
static const char *
stripwhitespace(char * buf)
{
  char * pointer = buf;
  char * revpointer;

  while (*pointer && isspace(*pointer))
    pointer ++;
  if (!*pointer)			/* if it is nothing, return NULL */
    return NULL;

  revpointer = pointer + strlen (pointer) - 1;

  /* strip whitespace from end. */
  while (revpointer > pointer && isspace(*revpointer))
    *(revpointer --) = 0;

  return pointer;
}


/**
 * Add hostname to machinelist if and only if 
 * the host does not already exist in the machine list.
 */
static linkedlist *
machinelist_lladd(linkedlist * machinelist,
		  const char * machinename)
{
  if (llmatch(machinelist, machinename)) /* if there is match, ignore */
    return machinelist;
  else
    return lladd(machinelist, machinename);
}


/*
 * Load machine list from netgroup (typically NIS)
 * Todo: Should perhaps remove duplicates.
 *
 * Code contributed from Petter Reinholdtsen on 22 Dec 2001.
 */
linkedlist*
read_machinenetgroup(linkedlist * machinelist,
                    const char * groupname)
{
  if (setnetgrent(groupname)) 
    {
      char *hostp=NULL, *userp=NULL, *domainp=NULL;
      
      while (getnetgrent(&hostp, &userp, &domainp))
	{
	  if (NULL != hostp)
	    machinelist = machinelist_lladd(machinelist, hostp);
	  
	}
      endnetgrent();
    } 
  else
    {
      fprintf(stderr, 
	      _("%s: Unknown netgroup %s.\n"),
	      PACKAGE, groupname);
    }
  return machinelist;
}

/**
 * read the machine list file from file.
 */
linkedlist* 
read_machinelist(linkedlist * machinelist, const char * listfile, const char*alternatelistfile)
{
  FILE * f;
  int bufferlen = 1024;  
  char * buf = malloc_with_error(bufferlen);
  
  if ((f = fopen (listfile, "r")) || ((NULL != alternatelistfile) && (f=fopen(alternatelistfile, "r"))))
    {
      while (-1 != getline (&buf, &bufferlen, f))
	{
	  const char * strippedstring = stripwhitespace(buf);
	  if (strippedstring)
	    machinelist=machinelist_lladd(machinelist,strippedstring); 
	}
      fclose(f);
    }
  else
    {
      if (alternatelistfile)
	fprintf (stderr, 
		 _("%s: File %s nor %s could not be opened for read\n"),
		 PACKAGE, listfile, alternatelistfile);
      else
	fprintf (stderr, 
		 _("%s: File %s could not be opened for read\n"), 
		 PACKAGE, listfile);
    }  
  free (buf);  
  return machinelist;  
}

int print_version(void)
{
  printf(_("Distributed Shell / Dancer\'s shell version %s \n"
	 "Copyright 2001,2002 Junichi Uekawa, \n"
	 "distributed under the terms and conditions of GPL version 2\n\n"),
	 VERSION);
  return 0;
}

int
print_help (void)
{
  print_version();
  printf(_(
	   "-v --verbose                   Verbose output\n"
	   "-q --quiet                     Quiet\n"	
	   "-M --show-machine-names        Prepend the host name on output\n"
	   "-i --duplicate-input           Duplicate input given to dsh\n"
	   "-b --bufsize                   Change buffer size used in input duplication\n"
	   "-u --user [username]           Provide login user name\n"
	   "-m --machine [machinename]     Execute on machine\n"
	   "-n --num-topology              How to divide the machines\n"
	   "-a --all                       Execute on all machines\n"
	   "-g --group [groupname]         Execute on group member\n"
	   "-f --file [file]               Use the file as list of machines\n"
	   "-r --remoteshell [shellname]   Execute using shell (rsh/ssh)\n"
	   "-o --remoteshellopt [option]   Option to give to shell \n"
	   "-h --help                      Give out this message\n"
	   "-w --wait-shell                Sequentially execute shell\n"
	   "-c --concurrent-shell          Execute shell concurrently\n"
	   "-V --version                   Give out version information\n\n")
 );
  return 0;
}


				/* load the configuration file. */
int
load_configfile(const char * dsh_conf)
{
  dshconfig * t;
  dshconfig_internal * line;
  FILE*f = fopen (dsh_conf, "r");  
  char * buf_a=NULL;
  char * buf_b=NULL;
  
  if(verbose_flag) printf(_("Loading config file %s\n"), dsh_conf);

  if (f && (t = open_dshconfig(f, '=')))
    {
      for (line = t->config; line; line=line->next)
	{
	  buf_a = line -> title;
	  buf_b = line -> data;

	  if(verbose_flag) printf(_(" Parameter %s is %s\n"), buf_a, buf_b);
	  if (!strcmp(buf_a, "remoteshell"))
	    {
		  if (verbose_flag) printf(_("Using %s as the remote shell\n"), buf_b);
		  remoteshell_command = strdup (buf_b);
	    }	      
	  else if (!strcmp(buf_a, "remoteshellopt"))
	    {
	      if (verbose_flag) printf(_("Adding [%s] to shell options\n"), buf_b);
	      remoteshell_command_opt_r = lladd (remoteshell_command_opt_r, buf_b);
	    }	  
	  else if (!strcmp(buf_a, "waitshell"))
	    {
	      wait_shell = atoi ( buf_b );		  
	      if (verbose_flag) printf(_("Setting wait-shell to  [%i]\n"), wait_shell);
	    }	      
	  else if (!strcmp(buf_a, "showmachinenames"))
	    {
	      pipe_option |= (atoi ( buf_b ) != 0 );
	      if (verbose_flag) printf(_("Setting pipe option to  [%i]\n"), pipe_option);
	    }	      
	  else if (!strcmp(buf_a, "verbose"))
	    {
	      verbose_flag = atoi ( buf_b );
	      if (verbose_flag) printf(_("Setting verbose to  [%i]\n"), verbose_flag);
	    }	      
	  else
	    {
	      fprintf (stderr, 
		       _("%s: unparsed configuration file line %s found in %s\n"),
		       PACKAGE, buf_a, dsh_conf);
	    }
	}

      free_dshconfig(t);
      fclose(f);
    }  
  return 0;
}

/** 
 * open /dev/null as stdin
 */
void 
open_devnull(void)
{
  int in = open ("/dev/null", O_RDONLY);
  dup2 (in, 0);
}


/**
 * Split the option string with "," as delimiter, and 
 * add the hostnames to the linked list.
 *
 * @return linked list with items added.
 */
static linkedlist*
split_machines_list_and_add_machines(linkedlist* machinelist, const char * optarg)
{
  char * s = strdup (optarg);
  char * next ;
  char * current = s;
  
  while ((next = strchr(current,',')))
    {
      *next = 0;
      machinelist = machinelist_lladd(machinelist, current);
      current = next + 1;
    }
  machinelist = machinelist_lladd(machinelist, current);
  
  free(s);
  return machinelist;
}

/**
 * Option parsing routine.
 *
 * Remember to update other places, such as execute_rsh_multiple routine when changing here.
 * @return 1 on failure.
 */
int
parse_options ( int ac, char ** av)
{
  int index_point;  
  int c;			/* option */
  linkedlist * machinelist = NULL;
  linkedlist * rshcommandline_r = NULL; /* command line for rsh in reverse order*/

#ifdef HAVE_GETOPT_LONG
  static struct option long_options[]=
  {
    {"verbose", no_argument, 0, 'v'},
    {"quiet", no_argument, 0, 'q'},
    {"show-machine-names", no_argument, 0, 'M'},
    {"duplicate-input", no_argument, 0, 'i'},
    {"bufsize", required_argument, 0, 'b'},

				/* machine-specification */
	{"user", required_argument, 0, 'u'},
    {"machine", required_argument, 0, 'm'},
    {"num-topology", required_argument, 0, 'n'},
    {"all", no_argument, 0, 'a' },
    {"group", required_argument, 0, 'g'},
    {"file", required_argument, 0, 'f'},

				/* rsh/ssh specification */
    {"remoteshell", required_argument, 0, 'r'},
    {"remoteshellopt", required_argument, 0, 'o'},

    {"help", no_argument, 0, 'h'},
    {"version", no_argument, 0, 'V'},
    {"wait-shell", no_argument, 0, 'w'},
	{"concurrent-shell", no_argument, 0, 'c'},
    {0,0,0,0}
  };
#else
#define getopt_long(a,b,c,d,e) getopt(a,b,c)
#endif  
  
  while((c = getopt_long (ac, av, "vqu:m:ar:f:g:hVcwo:Mn:ib:", long_options, &index_point)) != -1)
    {
      switch (c)
	{
	case 'a':
	  if (verbose_flag) printf (_("Adding all machines to the list\n"));
	  {
	    char * buf;
	    asprintf(&buf, "%s/.dsh/machines.list", getenv("HOME"));	    
	    machinelist = read_machinelist(machinelist, buf, DSHCONFDIR"/machines.list");
	    free (buf);	    
	  }	  
	  break;	  
	case 'g':
	  {
            if ('@' == *optarg)
              {			/* using libc call for using netgroup. */
                /* +1 to skip @ */
		if (verbose_flag) printf (_("Adding netgroup %s to the list\n"), optarg + 1);
                machinelist = read_machinenetgroup(machinelist, optarg+1);
              }
            else
              {			/* using dsh's own method. */
		char * buf1, *buf2;
		if (verbose_flag) printf (_("Adding group %s to the list\n"), optarg);
		asprintf(&buf1, DSHCONFDIR"/group/%s", optarg);
		asprintf(&buf2, "%s/.dsh/group/%s", getenv("HOME"), optarg);
		machinelist = read_machinelist (machinelist, buf2, buf1); 
		free(buf1);free(buf2);
	      }
	  }
	  break;	  
	case 'f':
	  if (verbose_flag) printf (_("Adding file %s to the list\n"), optarg);
	  machinelist = read_machinelist (machinelist, optarg, NULL); 
	  break;	  
	case 'v':
	  if (verbose_flag) printf (_("Verbose flag on\n"));
	  verbose_flag=1;	  
	  break;
	case 'q':
	  if (verbose_flag) printf (_("Verbose flag off\n"));
	  verbose_flag=0; 
	  break;
	case 'M':
	  if (verbose_flag) printf (_("Show machine names on output\n"));
	  pipe_option |= PIPE_OPTION_OUTPUT;
	  break;	  
	case 'i':
	  if (verbose_flag) printf (_("Duplicate input to all processes\n"));
	  pipe_option |= PIPE_OPTION_INPUT;
	  break;	  
	case 'b':
	  if (verbose_flag) printf (_("Buffer size used for dupliation\n"));
	  buffer_size = atoi(optarg);	  
	  break;
	case 'u':
		if (verbose_flag) printf (_("Adding user %s for login\n"), optarg);
	  login_user = strdup(optarg);
	  break;
	case 'm':
	  if (verbose_flag) printf (_("Adding machine %s to list\n"), optarg);
	  machinelist = split_machines_list_and_add_machines (machinelist, optarg);
	  break;
	case 'n':
	  if (verbose_flag) printf (_("Topology number set to %s\n"), optarg);
	  num_topology = atoi (optarg);	  
	  break;	  
	case 'h':
	  print_help();
	  exit(1);
	  break;	  
	case 'V':
	  print_version();
	  exit(1);
	  break;	  
	case 'r':		/* specify the shell command */
	  if (verbose_flag) printf(_("Using %s as the remote shell\n"), optarg);
	  remoteshell_command = strdup (optarg);
	  break;	  
	case 'o':		/* specify the shell command options */
	  if (verbose_flag) printf(_("Adding [%s] to shell options\n"), optarg);
	  remoteshell_command_opt_r = lladd (remoteshell_command_opt_r, optarg);
	  break;	  
	case 'w':		/* wait shell */
	  if (verbose_flag) printf (_("Wait for shell to finish executing\n"));
	  wait_shell=1;
	  break;
	case 'c':		/* concurrent shell */
	  if (verbose_flag) printf (_("Do not wait for shell to finish\n"));
	  wait_shell=0;
	  break;
	default:
	  if (verbose_flag) printf (_("Unhandled option\n"));
	  break;	  
	}
    }

  if (!machinelist)
    {
      fprintf (stderr, _("%s: no machine specified\n"), PACKAGE);
      return 1;
    }

  {				/* generate the command line for remote. */
    int i;
    for (i=optind; i < ac; ++i)
      {
	rshcommandline_r = lladd(rshcommandline_r, av[i]);
      }
  }
  
  /* sanity checking */
  if ((pipe_option & PIPE_OPTION_INPUT) && wait_shell )
    {
      fprintf (stderr, _("%s: Input duplication and concurrent shell need to be specified together\n"), PACKAGE);
      return 1;
    }
  

  if (!(pipe_option & PIPE_OPTION_INPUT))
    open_devnull();		/* open /dev/null if no input pipe is required */

				/* actually execute the code. */
  return do_shell(machinelist, rshcommandline_r);
}
